feat: use nix flake, std, tullia and cicero

jbgi opened the pull request:
std-based alternative take to #175.

hamishmack commented:
This looks really good.  Is there a way to pass evalSystem to everything?   I wanted to see if I could run `nix flake show` on macOS without using remote builder:

I tried the following, but I think the cross compiled versions of the project are still missing the `evalSystem` somehow.

```
diff --git a/cells/_automation/hydraJobs.nix b/cells/_automation/hydraJobs.nix
index 5d101b8..ceb7d5c 100644
--- a/cells/_automation/hydraJobs.nix
+++ b/cells/_automation/hydraJobs.nix
@@ -4,7 +4,7 @@
 }: let
   inherit (inputs.nixpkgs) lib;
   inherit (inputs.cells.prelude.library) project extra-compilers;
-  evalSystem = "x86_64-linux";
+  evalSystem = "x86_64-darwin";

   ciProject = project.appendModule {
     inherit evalSystem;
diff --git a/cells/prelude/library.nix b/cells/prelude/library.nix
index d949a48..9ed335a 100644
--- a/cells/prelude/library.nix
+++ b/cells/prelude/library.nix
@@ -41,6 +41,7 @@
     name = "cardano-prelude";
     inherit src;
     compiler-nix-name = lib.mkDefault default-compiler;
+    evalSystem = "x86_64-darwin";
     # Cross compilation support:
     crossPlatforms = p:
       lib.optionals pkgs.stdenv.hostPlatform.isLinux (
@@ -59,7 +60,7 @@
       # deduce package names from the cabal project to avoid hard-coding them:
       projectPackageNames = builtins.attrNames (haskellLib.selectProjectPackages
         (haskell-nix.cabalProject' {
-          inherit (config) name src compiler-nix-name cabalProjectLocal;
+          inherit (config) name src compiler-nix-name evalSystem cabalProjectLocal;
         })
         .hsPkgs);
     in [
```

jbgi commented:
@hamishmack right, forgot to pass evalSystem here: https://github.com/input-output-hk/cardano-prelude/pull/176/commits/bb6e26a8467751c64986dd9ea84446a4e2f50954
nix flake show should work now (with your local modifications).

dermetfan commented on cells/prelude/apps.nix:
```diff
@@ -0,0 +1,8 @@
+  inputs,
+  cell,
+}:
+# TODO: uniformize between "/", ":" and "__" seperators (first two have issues, with std, cicero, completions).
```
This should be fixed with latest tullia

```suggestion
```

jbgi commented:
@dermetfan tried your suggestion in last commit, but it is still failing:
```
$ tullia run ci/hydraJobs.x86_64-linux.roots
...
[✗] error  ci/hydraJobs.x86_64-linux.roots                                                                9m11.235325358s
warning: Git tree '/home/jbgi/Dev/iohk/cardano-prelude' is dirty                                                         
warning: Using saved setting for 'allow-import-from-derivation = true' from ~/.local/share/nix/trusted-settings.json.    
warning: Using saved setting for 'extra-substituters = https://cache.iog.io' from ~/.local/share/nix/trusted-settings.jso
n.                                                                                                                       
warning: Using saved setting for 'extra-trusted-public-keys = hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ='
from ~/.local/share/nix/trusted-settings.json.                                                                           
error (ignored): error: SQLite database '/home/jbgi/.cache/nix/eval-cache-                                               
v4/1553df62b97b634bce22160eed0bb4a05e9e574b5e5e752ad9c22245c16fd86e.sqlite' is busy                                      
error: flake 'git+file:///home/jbgi/Dev/iohk/cardano-prelude' does not provide attribute 'packages.x86_64-               
linux.tullia.x86_64-linux.task.ci/hydraJobs.x86_64-linux.roots.nsjail.run', 'legacyPackages.x86_64-linux.tullia.x86_64-  
linux.task.ci/hydraJobs.x86_64-linux.roots.nsjail.run' or 'tullia.x86_64-linux.task.ci/hydraJobs.x86_64-                 
linux.roots.nsjail.run'                                                                                                  
Failed to run /nix/store/i7zh5x24vsbv471zk7vadkgsa55qianz-nix-2.11.0/bin/nix build --json --no-link .#tullia.x86_64-     
linux.task.ci/hydraJobs.x86_64-linux.roots.nsjail.run: exit status 1
```

Also another issue is that even though I specify only one task, `tullia` execute all tasks... (almost crashing my machine: is there a way to limit concurrent jobs?).

dermetfan commented:
There was an escaping issue in tullia. Fixed that and updated the tullia flake input.

> is there a way to limit concurrent jobs?

I added a tullia task for each hydra job so you can now run them separately like this (without `ci/` prefix):

```
nix run github:input-output-hk/tullia -- run hydraJobs.x86_64-linux.ghc924.plan-nix
```

jbgi commented:
I'm now getting 
```
$ nix run github:input-output-hk/tullia -- run x86_64-linux._automation.ciJobs.ghc924.plan-nix
warning: Using saved setting for 'allow-import-from-derivation = true' from ~/.local/share/nix/trusted-settings.json.
warning: Using saved setting for 'extra-substituters = https://cache.iog.io' from ~/.local/share/nix/trusted-settings.json.
warning: Using saved setting for 'extra-trusted-public-keys = hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=' from ~/.local/share/nix/trusted-settings.json.
[✗] error  x86_64-linux._automation.ciJobs.ghc924.plan-nix                                                  12.778438199s
[E][2022-10-17T14:55:32+0000][1] void subproc::subprocNewProc(nsjconf_t*, int, int, int, int, int)():205                 
execve('/nix/store/zspimn40hjy77c6bdmk24lwam87zwmy2-x86_64-linux._automation.ciJobs.ghc924.plan-nix/bin/'x86_64-         
linux._automation.ciJobs.ghc924.plan-nix'') failed: No such file or directory                                            
[F][2022-10-17T14:55:32+0000][1] bool subproc::runChild(nsjconf_t*, int, int, int, int)():448 Launching child process    
failed                                                                                                                   
[W][2022-10-17T16:55:32+0200][802842] bool subproc::runChild(nsjconf_t*, int, int, int, int)():478 Received error message
from the child process before it has been executed                                                                       
[E][2022-10-17T16:55:32+0200][802842] int nsjail::standaloneMode(nsjconf_t*)():256 Couldn't launch the child process     
Failed to run /nix/store/6cka89h4lly1m068a2lv8i99gvp8a0x6-x86_64-linux._automation.ciJobs.ghc924.plan-nix-nsjail/bin/x86_
64-                                                                                                                      
linux._automation.ciJobs.ghc924.plan-nix-nsjail: exit status 255                                                         
4:55PM FTL starting supervisor error="running x86_64-linux._automation.ciJobs.ghc924.plan-nix: Failed to run /nix/store/6cka89h4lly1m068a2lv8i99gvp8a0x6-x86_64-linux._automation.ciJobs.ghc924.plan-nix-nsjail/bin/x86_64-linux._automation.ciJobs.ghc924.plan-nix-nsjail: exit status 255"
```

dermetfan commented:
That's just nsjail having trouble with cgroups when it tries to actually run the task (not yet well tested). Evaluation and building worked fine so it works [on cicero](https://cicero.ci.iog.io/run/7de95639-3252-4188-bc04-a4a0b0fd3dde), as for cicero the nomad + oci runtime is used instead of nsjail.

michaelpj commented on .envrc:
```diff
@@ -0,0 +1,8 @@
+DEVSHELL_TARGET=${DEVSHELL_TARGET:-default}
+
+source "$(nix eval .#__std.direnv_lib)"
+use std nix //_automation/devshells:${DEVSHELL_TARGET}
```
why don't we just say `use flake`?

michaelpj commented on cells/_automation/ciJobs.nix:
```diff
@@ -0,0 +1,25 @@
+  inherit (inputs.cells.prelude.library) project extra-compilers;
+
+  ciProject = project.appendModule {
+    flake.variants =
```
have I ever mentioned how much I like the "just use functions" approach? :D anyway, seems fine

michaelpj commented on cells/_automation/ciJobs.nix:
```diff
@@ -0,0 +1,25 @@
+  # path prefixes of jobs that should not be required for CI to pass (end up in `nonrequired`), empty for now:
+  nonRequiredPaths = map lib.hasPrefix [];
+in
+  nixpkgs.callPackages iohk-nix.utils.ciJobsAggregates {inherit ciJobs nonRequiredPaths;}
```
is this new? should we be using it in other places?

jbgi replied:
yeah I have been duplicating the logic to a few project, so upstreamed to iohk-nix today.

michaelpj commented on cells/_automation/devshells.nix:
```diff
@@ -0,0 +1,33 @@
+        {
+          package =
+            inputs.tullia.tullia.apps.tullia
+            // {meta.description = "CLI for Cicero tasks and actions";};
```
Can we fix this upstream?

jbgi replied:
right, forgot to do it. https://github.com/input-output-hk/tullia/pull/7

dermetfan replied:
Merged :heavy_check_mark:

michaelpj commented on cells/_automation/nixago.nix:
```diff
@@ -0,0 +1,55 @@
+in {
+  treefmt = presets.nixago.treefmt {
+    configData.formatter = {
+      prettier = {
```
do we actually need this?

michaelpj commented on cells/_automation/nixago.nix:
```diff
@@ -0,0 +1,55 @@
+  };
+
+  editorconfig = presets.nixago.editorconfig {
+    configData = {
```
Isn't this duplicating the checked in config file?

jbgi replied:
the checked-in config file is generated by this upon entering devshell.

michaelpj replied:
I think I still really don't get the point of `nixago`. Why are we checking it in if it's generated? or if it's checked in, why do we need to generate it?

jbgi replied:
> Why are we checking it in if it's generated?

Not sure (cc/ @blaggacao) but I believe for online editors, and maybe also for users that did not use `direnv`/`nix develop` before opening their editor.

blaggacao replied:
I didn't see this comment before. Sry.

Sometimes you may enter an editor without previously having entered the devshell.

I thought it was still good to have editorconfig in place so that the basic formatting is respected.

Now that I think about it I may have simplified and assumed that this file is loaded upon _starting_ the editor. But now I realize that it doesn't make sense. It must be loaded every time opening a file.

So actually my initial reasoning here was flawed.

You can change this relatively easily with setting `hook.mode = "link";`. Might be worth a try. You'd also have to remove the file from git tracking manually first, since link mode registers them in gitignore.

michaelpj commented on cells/_automation/tasks.nix:
```diff
@@ -0,0 +1,50 @@
+  mkJobTask = flakeOutputTask: {config, ...}: {
+    imports = [common flakeOutputTask];
+
+    # some hydra jobs run NixOS tests
```
none of these do!

michaelpj commented on cells/prelude/library.nix:
```diff
@@ -0,0 +1,92 @@
+  extra-compilers = ["ghc924"];
+
+  # We clean-up src to avoid rebuild for unrelated changes:
+  src = haskellLib.cleanSourceWith {
```
why doesn't `cleanGit` do what you want?

jbgi replied:
yeah removing a few filters that were redundant with flake builtin-in filtering (~ `cleanGit`).

michaelpj commented on release.nix:
```diff
@@ -6,65 +6,48 @@
-, projectArgs ? {
-    config = { allowUnfree = false; inHydra = true; };
-    gitrev = cardano-prelude.rev;
+{
```
should we just bin this if we're switching to cicero or do we want to keep both for now?

jbgi replied:
I think until we get mac agents for cicero we might want to keep both.

michaelpj commented on cells/prelude/library.nix:
```diff
@@ -0,0 +1,92 @@
+  };
+
+  # provide easy access to components, alternative compilers and a devshell module.
+  projectOverlays = with haskellLib.projectOverlays; [
```
I _really_ find this stuff way harder to think about than a simple function `CompilerName -> (Project, Shell)` or whatever, but :shrug:

michaelpj commented:
On reflection, a high-level thought. This PR contains hundreds of lines of nix code providing lots of features, but explains nothing about how to use it or work with it. To put it another way: `cardano-prelude` needs a `CONTRIBUTING`, and if we're going to add all this stuff it _desperately_ needs one. What is all this stuff? What is lefthook? What is nixago? What's up with this editorconfig thing? How do I get a shell? How do I update things? What's the CI setup? How do I change the compilers it's testing against? At the moment nobody will know any of this except @jbgi 

Yes, that's a lot to write about, but this PR _adds a lot of stuff_. If we want to keep it minimal instead that would be fine but if we're going to go big IMO we _have_ to document things.

jbgi commented:
Sure. Most std stuff can be quickly documented with links to the appropriate [std doc](https://std.divnix.com/index.html) page.

michaelpj commented:
I think that's good for maintainers of the nix code, but we also need to document things in a form that's comprehensible by people working on the Haskell code, hence CONTRIBUTING.

michaelpj commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+#### Available tools
+
+Available tools (`cabal`, `ghcid`, `hlint`, `weeder`, etc.) are displayed upon enterring the development shell. Type `menu` to see them again.
```
```suggestion
Available tools (`cabal`, `ghcid`, `hlint`, `weeder`, etc.) are displayed upon entering the development shell. Type `menu` to see them again.
```

michaelpj commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,119 @@
+
+```console
+$ cabal update
+$ cabal build cardano-prelude:lib:cardano-prelude
+$ cabal run cardano-prelude-test:test:prelude-tests
````
```suggestion
$ cabal build all
$ cabal test all
```

michaelpj commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+#### Commits
+
+1.  Format the source tree with `treefmt` (which use [`ormolu`](https://ormolu-live.tweag.io/))
```
how do I do this? is that the command to run?

michaelpj replied:
Have we agreed with the maintainers of this project that they want to use ormolu?

jbgi replied:
@newhoggy, @erikd are you ok to enforce formatting with https://github.com/tweag/ormolu?
(used in cardano-ledger, cardano-world, maybe other)
An alternative, more configurable is https://github.com/fourmolu/fourmolu

jbgi replied:
switched to fourmolu as it provide some work-around to ormulu default config, using same config as in cardano-node.

michaelpj commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+#### Commits
+
+1.  Format the source tree with `treefmt` (which use [`ormolu`](https://ormolu-live.tweag.io/))
+2.  Use [conventionnal](https://www.conventionalcommits.org/) commit message (`<type>[(<scope>)]: <short description>]). Available commit types are from [std presets](https://std.divnix.com/reference/presets/nixago/conform.html).
```
have we agreed with the maintainers of this project that they want to do this?

jbgi replied:
@newhoggy, @erikd are you ok with enforcing [conventional](https://www.conventionalcommits.org/) commit messages?

michaelpj commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,119 @@
+$ nix eval --raw .#$system.prelude.library.last-index-state
+```
+
+and copy the output date as `index-state` value in `cabal.project`.
````
I think this is backwards IMO: usually what I do is to update the `index-state` to whatever I want, and then just update `hackage.nix`. No need to ask hackage.nix for the latest index state or anything.

michaelpj commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,119 @@
+If you need to update [cabal.project](./cabal.project) `index-state`, you need to first update the nix pin for hackage:
+
+```console
+$ nix flake lock --update-input haskellNix
````
```suggestion
$ nix flake lock --update-input hackage
```

michaelpj commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+#### Commits
+
+1.  Format the source tree with `treefmt` (which use [`ormolu`](https://ormolu-live.tweag.io/))
```
What is this a list of?

michaelpj commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+(the std/nix build will use default compiler).
+
+#### Commits
```
missing editorconfig? Is that checked?

jbgi replied:
it is not checked, some editor may use it for tailor some formatting for this project, only `treefmt` not doing any change is what is checked (for now only by lefthook, should be added to CI too).

michaelpj commented on cells/prelude/devshellProfiles.nix:
```diff
@@ -0,0 +1,47 @@
+      {
+        package = haskell-nix.tool compiler-nix-name "ghcid" {
+          version = "latest";
+          index-state = last-index-state;
```
if you don't pass this argument `haskell.nix` will just use the latest index-state. I think the whole `last-index-state` thing is unnecessary.

jbgi replied:
it is unnecessary except that being explicit about it remove a warning (repeated of each tool).

michaelpj commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+# Contributing
```
:medal_military: :100: :1st_place_medal: :rocket:

michaelpj reviewed (commented):
CONTRIBUTING is great, thank you.

jbgi commented:
squashed commits, removed conventional commit enforcing, switch formatting to use fourmulu (same config as in `cardano-ledger`).

Should be ready to merge.

dermetfan commented:
I pushed a fix for the latest tullia version, maybe squash again @jbgi (didn't want to squash and force push to not confuse anybody)

dermetfan commented on bors.toml:
```diff
@@ -1,5 +1,6 @@
 status = [
-  "ci/hydra:Cardano:cardano-prelude:required",
+  "ci",
+  "ci (x86_64-linux)"
```
```suggestion
  "ci"
```

With the latest tullia version `ci` can only pass if `ci (x86_64-linux)` and all other available platforms also pass so just `ci` is sufficient if you want to require all platforms to pass. If you only want to require linux, list only `ci (x86_64-linux)` without `ci`.

dermetfan commented on cells/_automation/tasks.nix:
```diff
@@ -13,6 +13,7 @@
       github.ci = {
         enable = config.actionRun.facts != {};
         repository = "input-output-hk/cardano-prelude";
+        remote = config.preset.github.lib.readRepository inputs.cells._cloud.library.actionCiInputName null;
```
This is the remote to clone from. We need to read it from the input fact in order to support PRs from forks.

coot commented:
@erikd, @newhoggy, @lehins are you ok with requiring fourmolu formatter?

lehins commented:
> are you ok with requiring fourmolu formatter?

@coot I encourage requiring it!

blaggacao commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,130 @@
+
+### Commit message
+
+Commit messages should ideally be of the form `<type>[(<scope>)]: <short description>]` ([conventionnal](https://www.conventionalcommits.org/) commit message). Available commit types are from [std presets](https://std.divnix.com/reference/presets/nixago/conform.html).
```
Heads up: In my experience, the default scopes (one scope per cell) is often times not enough. So as this is being trained into muscle memory and used by people, I'd suggest _at least_ a refinement on the scopes.

blaggacao reviewed (commented):
For the formatter-only changes this will be handy: https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view

Please let me know or open an upstream ticket if you need anything from Standard. Even the ever so slightest improvement ideas are welcome!

jbgi commented:
bors r+
